contracts/README.md

```diff
@@ -3,15 +3,396 @@
+    function getResolver(string calldata label) external view returns (address);
+}
+```
+
+#### `RegistryDatastore` - Singleton Storage
+[src/common/RegistryDatastore.sol](src/common/RegistryDatastore.sol)
+
+Universal storage contract for all registries. Reduces storage proof complexity for L2 resolution.
+
+**Key Functions**:
+- `getEntry(address registry, uint256 tokenId)`: Fetch subregistry and resolver
+- `setSubregistry(uint256 tokenId, address subregistry)`: Update subregistry (caller must be registry)
+- `setResolver(uint256 tokenId, address resolver)`: Update resolver
+
+**Storage Structure**:
+```solidity
+struct Entry {
+    address subregistry;  // Registry for subdomains
+    address resolver;     // Resolver for this name
+    uint64 flags;         // Implementation-specific data (e.g., expiry)
+}
+```
+
+#### `PermissionedRegistry` - Standard Implementation
+[src/common/registry/PermissionedRegistry.sol](src/common/registry/PermissionedRegistry.sol)
+
+Feature-complete registry with role-based access control:
+- ERC1155 NFT for subdomains
+- Enhanced Access Control with 32 roles
+- Token regeneration support
+- Expiry management
+- Metadata support (name, description, image)
+
+#### `ERC1155Singleton` - Gas-Optimized NFT
+[src/common/ERC1155Singleton.sol](src/common/ERC1155Singleton.sol)
+
+Modified ERC1155 allowing only one token per ID:
+- Saves gas by omitting balance tracking
+- Provides `ownerOf(uint256 id)` like ERC721
+- Emits transfer events for indexing
+
+### L2 Components
+
+#### `ETHRegistrar` - .eth Name Registration
+[src/L2/registrar/ETHRegistrar.sol](src/L2/registrar/ETHRegistrar.sol)
+
+Handles .eth second-level domain registrations on L2:
+
+**Features**:
+- Commit-reveal registration (frontrunning protection)
+- Configurable pricing oracle
+- Multi-token payment support (ETH, USDC, etc.)
+- Minimum registration duration
+- Referral system
+
+**Registration Flow**:
+```solidity
+// 1. Commit to name + secret
+bytes32 commitment = keccak256(abi.encode(
+    label, owner, secret, subregistry, resolver, duration, referrer
+));
+registrar.commit(commitment);
+
+// 2. Wait MIN_COMMITMENT_AGE (prevent frontrunning)
+// 3. Complete registration
+registrar.register{value: price}(
+    label, owner, duration, secret, subregistry, resolver, referrer, paymentToken
+);
+```
+
+#### `StandardRentPriceOracle` - Dynamic Pricing
+[src/L2/registrar/StandardRentPriceOracle.sol](src/L2/registrar/StandardRentPriceOracle.sol)
+
+Length-based pricing with premium decay for expired names:
+- Shorter names cost more
+- Premium starts at 100% of annual rent and decays to 0% over 21 days
+- Fiat-pegged via Chainlink oracles
+- Multi-token support
+
+#### `L2EjectionController` - Name Migration
+[src/L2/bridge/L2EjectionController.sol](src/L2/bridge/L2EjectionController.sol)
+
+Manages ejection of names from L2 to L1:
+```solidity
+// Eject name to L1 for enhanced security
+ejectionController.initiateEjection(
```

Comment:
There is no `initiateEjection`. Ejection controllers are now known as bridge controllers and - for both L1 and L2 - operate by having name NFTs transferred to them.